private/bufpkg/bufapimodule/module_reader.go

```diff
@@ -37,35 +42,108 @@ func newModuleReader(
 	)
+	module, manifest, blobs, err := m.download(ctx, modulePin)
 	if err != nil {
-		if connect.CodeOf(err) == connect.CodeNotFound {
-			// Required by ModuleReader interface spec
-			return nil, storage.NewErrNotExist(modulePin.String())
-		}
 		return nil, err
 	}
-	moduleIdentity, err := bufmoduleref.NewModuleIdentity(
+	if manifest != nil {
+		// prefer and use manifest and blobs
+		return moduleFromBlobs(ctx, manifest, blobs, identityAndCommitOpt)
+	} else if module != nil {
+		// build proto module instead
+		return bufmodule.NewModuleForProto(ctx, module, identityAndCommitOpt)
+	}
+	// funny, success without a module to build
+	return nil, errors.New("no module in response")
+}
+
+func (m *moduleReader) download(
+	ctx context.Context,
+	modulePin bufmoduleref.ModulePin,
+) (
+	module *modulev1alpha1.Module,
+	manifest *modulev1alpha1.Blob,
+	blobs []*modulev1alpha1.Blob,
+	err error,
+) {
+	downloadService, err := m.downloadServiceProvider.NewDownloadService(
+		ctx,
 		modulePin.Remote(),
+	)
+	if err != nil {
+		return nil, nil, nil, err
+	}
+	module, manifest, blobs, err = downloadService.Download(
+		ctx,
 		modulePin.Owner(),
 		modulePin.Repository(),
+		modulePin.Commit(),
+	)
+	if err != nil && connect.CodeOf(err) == connect.CodeNotFound {
+		// Required by ModuleReader interface spec
+		err = storage.NewErrNotExist(modulePin.String())
+	}
+	return module, manifest, blobs, err
+}
+
+// moduleFromBlobs builds a module from a manifest blob and a set of other
+// blobs, provided in protobuf form. It returns a module built from a bucket.
+func moduleFromBlobs(
+	ctx context.Context,
+	manifestBlob *modulev1alpha1.Blob,
+	blobs []*modulev1alpha1.Blob,
+	options ...bufmodule.ModuleOption,
+) (bufmodule.Module, error) {
+	if _, err := manifest.FromProtoBlob(manifestBlob); err != nil {
+		return nil, fmt.Errorf("invalid manifest: %w", err)
+	}
+	parsedManifest, err := manifest.NewFromReader(
+		bytes.NewReader(manifestBlob.Content),
 	)
 	if err != nil {
-		return nil, err
+		return nil, fmt.Errorf("parse manifest content: %w", err)
+	}
+	var memBlobs []manifest.Blob
+	for i, modBlob := range blobs {
+		memBlob, err := manifest.FromProtoBlob(modBlob)
+		if err != nil {
+			return nil, fmt.Errorf("invalid blob at index %d: %w", i, err)
+		}
+		memBlobs = append(memBlobs, memBlob)
 	}
-	return bufmodule.NewModuleForProto(
-		ctx, module,
-		bufmodule.ModuleWithModuleIdentityAndCommit(moduleIdentity, modulePin.Commit()),
+	blobSet, err := manifest.NewBlobSet(
+		ctx,
+		memBlobs,
+		manifest.BlobSetWithContentValidation(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("invalid blobs: %w", err)
+	}
+	manifestBucket, err := manifest.NewBucket(
+		*parsedManifest,
+		*blobSet,
+		manifest.BucketWithAllManifestBlobsValidation(),
+		manifest.BucketWithNoExtraBlobsValidation(),
+	)
+	if err != nil {
+		return nil, fmt.Errorf("new manifest bucket: %w", err)
+	}
+	return bufmodule.NewModuleForBucket(
+		ctx,
+		manifestBucket,
+		options...,
```

Comment:
That is a thin wrapper on bufmodulebuild.moduleBucketBuilder. It can take a bucket that's not constrained to just the content needed for a module. Its intent is to be provided a bucket with a large set of files which a config is located, proto files found, and mapped appropriately into a cleaned read bucket. It calls `bufmodule.NewModuleForBucket` after all that.

I'm skipping all that to be more strict, something we can change later if we have to consume more complicated manifests.